### PR TITLE
IVT computation app

### LINKED ISSUE #391: IVT calculation app for HighResMIP
We need an app for calculating integrated vapor transport from the HighResMIP models.

It will need to accept files containing both 4D (time, lev, lat ,lon) and 3D (time, lat, lon) variables:

- 4D variables
  - humidity Q
  - zonal wind U
  - meridional wind V
- 3D variables
  - surface humidity QSFC
  - surface zonal wind USFC
  - surface meridional wind VSFC
  - surface pressure PS, or
    - surface sea-level pressure
    - surface elevation (a 2D [lat, lon] field)

It will output netCDF files containing IVT, a 3D field that contains the vertical integral of horizontal moisture transport.

A typical run command should be something like:
```bash
teca_ivt_integral \
    --q_file_regex QFILEREGEX \
    --u_file_regex UFILEREGEX \
    --v_file_regex VFILEREGEX \
    --qsfc_file_regex QSFCFILEREGEX \
    --usfc_file_regex USFCFILEREGEX \
    --vsfc_file_regex VSFCFILEREGEX \
    --ps_file_regex PSFILEREGEX
    --q_var_name QVARNAME \
    --u_var_name UVARNAME \
    --v_var_name VVARNAME \
    --ps_var_name PSVARNAME \
    --output_ivt_var_name IVTVARNAME \
    --output_file_name OUTPUTFILETEMPLATE \
```

The pipeline will be something like:
```
CF Reader --> Normalize Coordinates --> Moisture Transport vector (Q*U, V*Q) --> Vertical Integral (uIVT, vIVT) --> l2_norm (IVT) --> CF writer
```

Depends on:

- #353 
- #390

### PR DESCRIPTION
**Pull request by elbashandy**
Resolves #391

### CODE REVIEW

File: apps/CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,9 @@ endif()
 teca_add_app(teca_bayesian_ar_detect LIBS ${teca_app_link}
     FEATURES ${TECA_HAS_BOOST} ${TECA_HAS_NETCDF} ${TECA_HAS_UDUNITS})
 
+teca_add_app(teca_ivt_computation LIBS ${teca_app_link}
```
Comment (burlen) [inline apps/CMakeLists.txt:22]: let's change the name to `teca_integrated_vapor_transport`

File: apps/teca_ivt_computation.cpp [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+            "index of reader that has the variable that contains the longitudinal component of the wind vector (1)")
+        ("wind_v_reader_index", value<unsigned int>()->default_value(1),
+            "index of reader that has the variable that contains the latitudinal component of the wind vector (1)")
+        ("specific_humidity_reader_index", value<unsigned int>()->default_value(1),
+            "index of reader that has the variable that contains the specific humidity (1)")
+        ("time_reader_index", value<unsigned int>()->default_value(1),
+            "index of reader that provides time axis (1)")
+        ("geometry_reader_index", value<unsigned int>()->default_value(1),
+            "index of reader the provides mesh geometry (1)")
+        ("point_arrays", value<std::vector<std::string>>()->multitoken(), "list of point centered arrays to write")
```
Comment (burlen) [inline apps/teca_ivt_computation.cpp:61]: not needed

**Review by burlen — CHANGES_REQUESTED** [review]
Looking good but I see a few things that should be changed.

**Comment by elbashandy** [comment]
@burlen Thanks for reviewing this. I think I covered all the required changes. Please give it a look when you have a chance.

File: apps/teca_cartesian_mesh_diff.cpp [hunk 3]
```diff
@@ -30,8 +30,8 @@ int main(int argc, char **argv)
         return -1;
     }
 
-    std::string test_file = argv[1];
-    std::string ref_file = argv[2];
+    std::string ref_file = argv[1];
+    std::string test_file = argv[2];
```
[the remark this answers is not in the document]
Comment (elbashandy) [reply apps/teca_cartesian_mesh_diff.cpp:34]: Yes I must have made a mistake while rebasing. Just checked, everything looks in place.

**Review by burlen — CHANGES_REQUESTED** [review]
a few changes to be made

File: test/apps/test_integrated_vapor_transport_app.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+${launcher} ${app_prefix}/teca_integrated_vapor_transport --readers                                                                       \
+    "r_0,${data_root}/HighResMIP/ECMWF-IFS-HR-SST-present/ua/ua_6hrPlevPt_ECMWF-IFS-HR_highresSST-present_r1i1p1f1_gr_1950-.*\\.nc,ua"    \
+    "r_1,${data_root}/HighResMIP/ECMWF-IFS-HR-SST-present/va/va_6hrPlevPt_ECMWF-IFS-HR_highresSST-present_r1i1p1f1_gr_1950-.*\\.nc,va"    \
+    "r_2,${data_root}/HighResMIP/ECMWF-IFS-HR-SST-present/hus/hus_6hrPlevPt_ECMWF-IFS-HR_highresSST-present_r1i1p1f1_gr_1950-.*\\.nc,hus" \
+    --steps_per_file 127 --verbose --output_file test_ivt_computation_app_output.nc
+
+# run the diff
+${app_prefix}/teca_cartesian_mesh_diff                        \
+    "${data_root}/test_ivt_computation_app_ref.bin"           \
+    "test_ivt_computation_app_output.*\.nc"
```
[the remark this answers is not in the document]
Comment (burlen) [reply test/apps/test_integrated_vapor_transport_app.sh:32]: confirmed. teca_cartesian_mesh diff doesn't use the index executive as a result an empty request is always made, hence step 0 is always requested and no arrays are being compared. this also explains why it didn't matter which order the reference and test datasets were passed on the command line. Added an issue for this #498

**Review by burlen — APPROVED** [review]
all set, but depends on #499. hold off until that is merged

**Pull request merged**